Feat/Datakit plugin (#1851)
Kong/developer.konghq.com: app/_kong_plugins/datakit/index.md

  - node: "`response`"
    inputs: "`body`, `headers`"
    outputs: none
    description: Updates properties of the outgoing client response
{% endtable %}
<!--vale off-->

#### Headers

The `headers` type produces and consumes maps from header names to their values:

* Keys are header names. Original header name casing is preserved for maximum
    compatibility.
* Values are strings if there is a single instance of a header or arrays of
    strings if there are multiple instances of the same header.

#### Query

The `query` type produces and consumes maps with key-value pairs representing
decoded URL query strings.

#### Body

The `service_request.body` and `response.body` inputs both accept any data type.
If the data is an object, it will automatically be JSON-encoded, and the
`Content-Type` header set to `application/json` (if not already set in the
`headers` input).

The `request.body` and `service_response.body` outputs have a similar behavior.
If the corresponding `Content-Type` header matches the JSON mime-type, the
`body` output is automatically JSON-decoded.

### `call` node

Send an HTTP request and retrieve the response.

> #### Inputs

Reviewer: ```suggestion
Inputs:
```

* `body`: Request body
* `headers`: Request headers
* `query`: Key-value pairs to encode as the request query string

> #### Outputs

Reviewer: ```suggestion
Outputs:
```

* `body`: The response body
* `headers`: The response headers
* `status`: The HTTP status code of the response

> #### Configuration attributes

Reviewer: ```suggestion
Configuration attributes:
```

depend on it). Multiple call nodes are executed concurrently when no dependency
order enforces it.

In this example, both `CALL_FOO` and `CALL_BAR` will be started as soon as
possible, and then Datakit will block until both have finished to run
`JOIN`:

```yaml
- name: CALL_FOO
  type: call
  url: https://example.com/foo

- name: CALL_BAR
  type: call
  url: https://example.com/bar

- name: JOIN
  type: jq
  jq: "."
  inputs:
    foo: CALL_FOO.body
    bar: CALL_BAR.body
```

#### Failure conditions

The `call` node fails execution if a network-level error is encountered or if
the endpoint returns a non-2xx status code. It will also fail if the endpoint
returns a JSON mime-type in the `Content-Type` header if the response body is
not valid JSON.

#### Limitations

Due to platform limitations, the `call` node can't be executed after proxying a
request, so attempting to configure the node using outputs from the upstream service
response will yield an error:

```yaml
- name: CALL
  type: call
  url: https://example.com/
  method: POST
  inputs:
    # dependency error!
    body: service_response.body
```

### `jq` node type

> Execution of a jq script for processing JSON. See the official

Reviewer: ```suggestion
The `jq` node executes a jq script for processing JSON. See the official
```
Not sure if this is accurate, but the original sentence is incomplete (missing subject)

```yaml
- name: STRING
  name: jq
  jq: |
    "my string"

- name: NUMBER
  name: jq
  jq: |
    54321

- name: BOOLEAN
  name: jq
  jq: |
    true

- name: OBJECT
  name: jq
  jq: |
    {
      a: 1,
      b: 2
    }
```

It's impossible for Datakit to know ahead of time what kind of data `jq` will
emit, so Datakit uses runtime checks when the output of `jq` is connected to
another node's input. It's important to carefully test and validate your Datakit
configurations to avoid this case:

```yaml
- name: HEADERS
  name: jq
  jq: |
    "oops, not an object/map"

- name: EXIT
  type: exit
  inputs:
    # this will cause Datakit to return a 500 error to the client when
    # encountered
    headers: HEADERS
```

This is also why the `jq` node doesn't allow explicitly referencing individual
fields with `outputs` at config-time:

```yaml
- name: HEADERS
  name: jq
  jq: |
    "this is completely opaque to Datakit"

  # Datakit will reject this configuration because it can't confirm that the
  # output of HEADERS is an object or has a `body` field
  outputs:
    body: EXIT.body

- name: EXIT
  type: exit
```

#### Configuration attributes

> * `jq`: the jq script to execute when the node is triggered.

Reviewer: ```suggestion
`jq`: the jq script to execute when the node is triggered.
```
